crates/matrix-sdk-crypto/src/session_manager/group_sessions/share_strategy.rs

```diff
@@ -209,6 +253,45 @@ pub(crate) async fn collect_session_recipients(
     Ok(CollectRecipientsResult { should_rotate, devices, withheld_devices })
 }
 
+// Check whether any of the devices currently in the discussion (outbound
+// session stores the devices it already got shared with) got deleted or
+// blacklisted in the meantime. If it's the case the session should be rotated.
+fn should_rotate_due_to_left_device(
+    user_id: &UserId,
+    future_recipients: &[DeviceData],
+    outbound: &OutboundGroupSession,
+) -> bool {
+    // Device IDs that should receive this session
+    let recipient_device_ids: BTreeSet<&DeviceId> =
+        future_recipients.iter().map(|d| d.device_id()).collect();
+
+    if let Some(shared) = outbound.shared_with_set.read().unwrap().get(user_id) {
+        // Devices that received this session
+        let shared: BTreeSet<OwnedDeviceId> = shared.keys().cloned().collect();
+        let shared: BTreeSet<&DeviceId> = shared.iter().map(|d| d.as_ref()).collect();
```

Comment:
Do we need to collect it twice?

Reply:
+1. I fixed another copy of this code recently in https://github.com/matrix-org/matrix-rust-sdk/commit/1e58c0382c6dcc3f888c39b5cb0d5cc0264aa9fe